lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,409 @@
+    <tarball_directory_name>/<tarball_name>
+    """
+    return os.path.join(tarball_directory_name(spec),
+                        tarball_name(spec, ext))
+
+
+def build_tarball(spec, outdir, force=False, rel=False, yes_to_all=False,
+                  key=None):
+    """
+    Build a tarball from given spec and put it into the directory structure
+    used at the mirror (following <tarball_directory_name>).
+    """
+    tarfile_name = tarball_name(spec, '.tar.gz')
+    tarfile_dir = join_path(outdir, "build_cache",
+                            tarball_directory_name(spec))
+    tarfile_path = join_path(tarfile_dir, tarfile_name)
+    mkdirp(tarfile_dir)
+    spackfile_path = os.path.join(
+        outdir, "build_cache", tarball_path_name(spec, '.spack'))
+    if os.path.exists(spackfile_path):
+        if force:
+            os.remove(spackfile_path)
+        else:
+            tty.warn("file exists, use -f to force overwrite: %s" %
+                     spackfile_path)
+            return
+
+    # need to copy the spec file so the build cache can be downloaded
+    # without concretizing with the current spack packages
+    # and preferences
+    spec_file = join_path(spec.prefix, ".spack", "spec.yaml")
+    specfile_name = tarball_name(spec, '.spec.yaml')
+    specfile_path = join_path(outdir, "build_cache", specfile_name)
+    indexfile_path = join_path(outdir, "build_cache", "index.html")
+    if os.path.exists(specfile_path):
+        if force:
+            os.remove(specfile_path)
+        else:
+            tty.warn("file exists, use -f to force overwrite: %s" %
+                     specfile_path)
+            return
+    shutil.copyfile(spec_file, specfile_path)
+
+    # create info for later relocation and create tar
+    write_buildinfo_file(spec)
+    if rel:
+        prelocate_package(spec)
+    with closing(tarfile.open(tarfile_path, 'w:gz')) as tar:
+        tar.add(name='%s' % spec.prefix, arcname='%s' %
+                os.path.basename(spec.prefix))
+
+    # Sign the packages if keys available
+    sign = True
+    if not has_gnupg2():
+        tty.warn('gpg2 is not available for signing.')
+        sign = False
+    else:
+        if key is None:
+            keys = Gpg.signing_keys()
+            if len(keys) == 1:
+                key = keys[0]
+            if len(keys) > 1:
+                tty.warn(
+                    'Multiple signing keys are available.' +
+                    ' please choose one with -k option.')
+                Gpg.list(False, True)
+                sign = False
+            if len(keys) == 0:
+                tty.warn('No signing keys are available.')
+                sign = False
+    if sign:
+        Gpg.sign(key, specfile_path, '%s.asc' % specfile_path)
+        Gpg.sign(key, tarfile_path, '%s.asc' % tarfile_path)
+    else:
+        if not yes_to_all:
+            raise RuntimeError('Not creating unsigned build cache.' +
+                               'Use spack buildcache create -y ' +
```

Comment:
The `+` isn't necessary either; Python concatenates literal strings in the source code.